[scroll-animations] update `AnimationEffectTiming` and `BasicEffectTiming` to use `CSSNumberishTime`

graouts opened the pull request:
#### 82e54eaf31dbcf921a25b93e0751869a876bd04f
<pre>
[scroll-animations] update `AnimationEffectTiming` and `BasicEffectTiming` to use `CSSNumberishTime`
<a href="https://bugs.webkit.org/show_bug.cgi?id=280823">https://bugs.webkit.org/show_bug.cgi?id=280823</a>
<a href="https://rdar.apple.com/137200964">rdar://137200964</a>

Reviewed by Cameron McCormack.

We refactor yet more code backing timing computations to `CSSNumberishTime` instead of `Seconds`,
paving the way to supporting percentage values for Scroll-driven Animations.

* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::durationAPIValue):
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getComputedTiming const):
* Source/WebCore/animation/AnimationEffect.h:
(WebCore::AnimationEffect::iterationDuration const):
(WebCore::AnimationEffect::activeDuration const):
(WebCore::AnimationEffect::endTime const):
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::updateComputedProperties):
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/BasicEffectTiming.h:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::CSSNumberishTime):
(WebCore::CSSNumberishTime::isInfinity const):
(WebCore::CSSNumberishTime::isZero const):
(WebCore::CSSNumberishTime::approximatelyEqualTo const):
(WebCore::CSSNumberishTime::operator/ const):
* Source/WebCore/animation/CSSNumberishTime.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer const):
* Source/WebCore/animation/KeyframeInterpolation.cpp:
(WebCore::transformProgressDuration):
(WebCore::KeyframeInterpolation::interpolateKeyframes const):
* Source/WebCore/animation/KeyframeInterpolation.h:
* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::invalidateDOMEvents):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::zeroTime const):
(WebCore::WebAnimation::effectEndTime const):
(WebCore::WebAnimation::finish):
(WebCore::WebAnimation::updateFinishedState):
(WebCore::WebAnimation::pause):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/284666@main">https://commits.webkit.org/284666@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41ee5f9e0ba5a21b232911eaf8cf7a3158b5723a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70166 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49569 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22929 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21332 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72283 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21184 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73232 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/45137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/60506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/41797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19701 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/18284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/75970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14391 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/75970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14431 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/60575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/75970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/11320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10718 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45374 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -661,11 +661,17 @@ auto WebAnimation::playState() const -> PlayState
 }
 
-Seconds WebAnimation::effectEndTime() const
+CSSNumberishTime WebAnimation::zeroTime() const
```
Should this be a static method?

graouts replied:
Right now, it could be. But in a future patch (watch this space) we will query the attached timeline to determine whether we should return a percentage value or a time value.

heycam commented on Source/WebCore/animation/KeyframeInterpolation.cpp:
```diff
@@ -121,7 +121,14 @@ const KeyframeInterpolation::KeyframeInterval KeyframeInterpolation::interpolati
 }
 
-void KeyframeInterpolation::interpolateKeyframes(Property property, const KeyframeInterval& interval, double iterationProgress, double currentIteration, Seconds iterationDuration, TimingFunction::Before before, const CompositionCallback& compositionCallback, const AccumulationCallback& accumulationCallback, const InterpolationCallback& interpolationCallback, const RequiresBlendingForAccumulativeI ...
+static double transformProgressDuration(const CSSNumberishTime& duration)
```
I don't really understand what this function is used for.

graouts replied:
Yeah… this is specific to the `CubicBezierFunction` and  type of `TimingFunction::transformProgress()` for the `UnitBezier` solver. I suppose we could pass down the `CSSNumberishTime` through here.

heycam commented on Source/WebCore/animation/CSSNumberishTime.cpp:
```diff
@@ -105,6 +133,18 @@ CSSNumberishTime CSSNumberishTime::operator-(const CSSNumberishTime& other) cons
+{
+    ASSERT(m_type == other.m_type);
+    return { m_type, m_value / other.m_value };
+}
```
Dimensionally these don't really make sense to me. `operator*` should return like a time^2 value, and the second one a unitless value. Can / should we be using some different types here?

graouts replied:
I don't think we actually need `operator*` as it stands. I agree that `operator/` should return a `double` and will make that change.

graouts replied:
And as it turns out the only call site for `operator/` would immediately use `toDouble()` for the returned value.

heycam commented on Source/WebCore/animation/CSSNumberishTime.h:
```diff
@@ -33,17 +33,23 @@ namespace WebCore {
 class CSSNumberishTime {
 public:
     CSSNumberishTime() = default;
+    WEBCORE_EXPORT CSSNumberishTime(std::optional<Seconds>, std::optional<double>);
```
Does this constructor have to be written like this (and potentially take invalid combinations of arguments) because of the generated IPC deserializer?

graouts replied:
That is indeed the reason why. The implementation will `ASSERT()` if the input is incorrect. We could write custom encoding/decoding code for that purpose. I agree this is not ideal and I will look into improving this as a follow-up.

heycam commented on Source/WebCore/animation/CSSNumberishTime.cpp:
```diff
@@ -145,7 +185,10 @@ bool CSSNumberishTime::operator>=(const CSSNumberishTime& other) const
+    if (m_type == Type::Time)
+        return std::abs(time()->microseconds() - other.time()->microseconds()) < timeEpsilon.microseconds();
+    return m_value == other.m_value;
 }
```
Nit: I generally don't like approximate equal checks to be done inside `operator==` because it's not obvious from the use of `==` that it's doing approximate checks.

graouts replied:
Yeah, this should be part of a custom method. I had it originally as an `equals()` method but I will rename to `approximatelyEqualTo()`.

graouts committed:
[scroll-animations] update `AnimationEffectTiming` and `BasicEffectTiming` to use `CSSNumberishTime`
https://bugs.webkit.org/show_bug.cgi?id=280823
rdar://137200964

Reviewed by Cameron McCormack.

We refactor yet more code backing timing computations to `CSSNumberishTime` instead of `Seconds`,
paving the way to supporting percentage values for Scroll-driven Animations.

* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::durationAPIValue):
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getComputedTiming const):
* Source/WebCore/animation/AnimationEffect.h:
(WebCore::AnimationEffect::iterationDuration const):
(WebCore::AnimationEffect::activeDuration const):
(WebCore::AnimationEffect::endTime const):
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::updateComputedProperties):
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/BasicEffectTiming.h:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::CSSNumberishTime):
(WebCore::CSSNumberishTime::isInfinity const):
(WebCore::CSSNumberishTime::isZero const):
(WebCore::CSSNumberishTime::approximatelyEqualTo const):
(WebCore::CSSNumberishTime::operator/ const):
* Source/WebCore/animation/CSSNumberishTime.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer const):
* Source/WebCore/animation/KeyframeInterpolation.cpp:
(WebCore::transformProgressDuration):
(WebCore::KeyframeInterpolation::interpolateKeyframes const):
* Source/WebCore/animation/KeyframeInterpolation.h:
* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::invalidateDOMEvents):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::zeroTime const):
(WebCore::WebAnimation::effectEndTime const):
(WebCore::WebAnimation::finish):
(WebCore::WebAnimation::updateFinishedState):
(WebCore::WebAnimation::pause):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/284666@main
